doc/manual/src/language/string-context.md

```diff
@@ -0,0 +1,145 @@
+# String context
+
+> **Note**
+>
+> This is an advanced topic.
+> The Nix language is designed to be used without the programmer knowing string contexts are.
+
+Strings in the Nix language are not just a sequences of characters like strings in other languages.
+They are actually pairs of sequences of characters and what is known as a *string context*.
+The string context is an (unordered) set of *string context elements*.
+
+The goal of string contexts are to collect non-string values combined with strings via
+[string concatenation](./operators.md#string-concatenation),
+[string interpolation](./string-interpolation.md),
+and similar operations.
+The idea is that a user can combine together values to create a build recipe without manually keeping track of where the "ingredients" come from, and then the Nix language does that bookkeeping implicitly to come up with the right derivation inputs.
+
+> In line with this goal, string contexts are *not* explicitly manipulated in idiomatic Nix code.
+> Strings with non-empty context are only concatenated and eventually passed to `builtins.derivation`.
+> "Regular" strings an empty context are the only ones inspected, e.g. compared with `==`.
+
+String context elements come in 3 forms:
+
+- [*constant*]{#string-context-element-constant}
+- [*output*]{#string-context-element-output}
+- [*derivation deep*]{#string-context-element-derivation-deep}
+
+*Constant* and *output* string contexts elements are just
+[deriving paths](@docroot@/glossary.md#gloss-deriving-path);
+those are just the names of the two kinds of deriving path.
+See the documentation on deriving paths for further details.
+
+*derivation deep* is an advanced feature intended to be used with the
+[`exportReferencesGraph`](./advanced-attributes.html#adv-attr-exportReferencesGraph)
+advanced derivation feature.
+A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:
+all its outputs, all the other derivations the given derivation depends on, and all their outputs too.
```

Comment:
I feel weird not given each item its own bullet point, however. Especially since we removed the `3`.

We could however really embrace the subgroup and say that a string context is either a derived path or a derivation deep, and let the user read about derived paths separately. Then it really would be 2 kinds, with 2 kinds within the first kind.

Reply:
```suggestion
- [deriving path]{#string-context-element-derived-path}

  A string context element of this type is a [deriving path](@docroot@/glossary.md#gloss-deriving-path).
  They can be one of the varieties *constant* or *output*, which correspond to the types of deriving path.
  
  <!-- put the two sections with examples here as a list with anchors on the items -->

- [*derivation deep*]{#string-context-element-derivation-deep}

  *derivation deep* is an advanced feature intended to be used with the
  [`exportReferencesGraph` derivation attribute](./advanced-attributes.html#adv-attr-exportReferencesGraph).

  A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:
  all its outputs, all the other derivations the given derivation depends on, and all the outputs of those.
```